src/jgikbase/test/idmapping/core/mapper_test.py

```diff
@@ -28,34 +28,82 @@ def fail_init(handlers, storage, expected):
     idm = IDMapper(set(), storage)
 
-    with raises(Exception) as got:
-        idm.create_namespace(None)
-    assert_exception_correct(got.value, TypeError('namespace_id cannot be None'))
+    as_ = AuthsourceID('foo')
+    t = Token('t')
+    n = NamespaceID('n')
 
+    fail_create_namespace(idm, None, t, n, TypeError('authsource_id cannot be None'))
+    fail_create_namespace(idm, as_, None, n, TypeError('token cannot be None'))
+    fail_create_namespace(idm, as_, t, None, TypeError('namespace_id cannot be None'))
 
-def test_add_user_to_namespace():
+
+def test_create_namespace_fail_no_authsource():
     storage = create_autospec(IDMappingStorage, spec_set=True, instance=True)
     handler = create_autospec(UserHandler, spec_set=True, instance=True)
+    handler.get_authsource_id.return_value = AuthsourceID('as')
 
-    handler.get_authsource_id.return_value = AuthsourceID('asone')
     idm = IDMapper(set([handler]), storage)
 
-    handler.is_valid_user.return_value = True
+    fail_create_namespace(idm, AuthsourceID('bs'), Token('t'), NamespaceID('n'),
+                          NoSuchAuthsourceError('bs'))
+
+
+def test_create_namespace_fail_not_admin():
+    storage = create_autospec(IDMappingStorage, spec_set=True, instance=True)
+    handler = create_autospec(UserHandler, spec_set=True, instance=True)
+    handler.get_authsource_id.return_value = AuthsourceID('as')
+
+    idm = IDMapper(set([handler]), storage)
+
+    handler.get_user.return_value = (User(AuthsourceID('as'), Username('foo')), False)
+
+    fail_create_namespace(idm, AuthsourceID('as'), Token('t'), NamespaceID('n'),
+                          UnauthorizedError('User as/foo is not a system administrator'))
 
-    idm.add_user_to_namespace(NamespaceID('ns1'), User(AuthsourceID('asone'), Username('u1')))
 
-    assert handler.is_valid_user.call_args_list == [((Username('u1'),), {})]
+def fail_create_namespace(idm, authsource, token, namespace_id, expected):
+    with raises(Exception) as got:
+        idm.create_namespace(authsource, token, namespace_id)
+    assert_exception_correct(got.value, expected)
+
+
+def test_add_user_to_namespace():
+    storage = create_autospec(IDMappingStorage, spec_set=True, instance=True)
+    handler1 = create_autospec(UserHandler, spec_set=True, instance=True)
+    handler2 = create_autospec(UserHandler, spec_set=True, instance=True)
+
+    handler1.get_authsource_id.return_value = AuthsourceID('asone')
+    handler2.get_authsource_id.return_value = AuthsourceID('astwo')
+    idm = IDMapper(set([handler1, handler2]), storage)
+
+    handler2.get_user.return_value = (User(AuthsourceID('astwo'), Username('foo')), True)
+    handler1.is_valid_user.return_value = True
+
+    idm.add_user_to_namespace(
+        AuthsourceID('astwo'),
+        Token('t'),
+        NamespaceID('ns1'),
+        User(AuthsourceID('asone'), Username('u1')))
+
+    assert handler2.get_user.call_args_list == [((Token('t'),), {})]
```

Comment:
You might find this article interesting: https://martinfowler.com/articles/mocksArentStubs.html